### ISSUE TITLE
How does Embedding Layer work?

### ISSUE THREAD

**Issue by amityaffliction** [fchollet/keras#3110]
I'm curious about how embedding layers work

Embedding layer is trained for objective function that user can specify. for example 'mse' can be passed as argument

Q1:
but when using 'mse'  which is (0.5) \* (target - output)^2  what is target and output when training embedding layer? and what is embedding vector that model generates?

here is what I understand. is it right?

for example input takes sentences in each word as index (ex [0,12, 4, 4, 1]) and embedding model(RNN) is trained to predict next word for example

input 0 -> ouput 12 
input 12 -> output 4 
.. and so on...
and error metric is 'mse'

is it right? where does embedding vector come from? 

I thought it as RNN hidden state activation but hidden state activation changes through context...

Q2:
is there Keras example that can be used as Mikolov Word2Vec model?

Q3:
sentence can be variable length sequence that consists of different number of words.
how can we train embedding layer with variable length sequence??? do we have to pad it with special token?
I think it seems inefficient

**Comment by jmhessel** [comment]
1. 'mse' can't really be passed as an argument to Embedding. 'mse' can be a loss function for a model that contains an embedding layer, however. Embedding layers map from indices to vectors, i.e. index 7 to vector [3.2,-.3,.52...]. The parameters of the model are the lookup table values.

2. Here's my implementation of skip-gram in keras. This has not been thoroughly tested, so use at your own discretion.

```python
def make_word2vec_model(num_ns, embedding_dim, num_nodes, seperate_context = True):
	'''
	num_ns: (int) number of negative samples
	embedding_dim: (int) embedding dimension
	num_words: (int) size of the vocabulary
	seperate_context: (bool) whether or not to use a seperate set of word embeddings
							for the context embeddings. Reccomended for text data to
							better fit the distributional hypothesis.

	creates a skipgram with negative sampling model. word_input is a (batch_size, 1)
	shaped tensor indicating the index of the center word in the sliding window. context_input
	is a (batch_size, num_ns+1) shaped tensor, where the first column are the indices
	of the positive samples, and the num_ns following columns are the indices of the negative
	samples. Labels should be a (batch_size, num_ns+1) shaped tensor, where the first column
	is ones (positive label) and the rest of the matrix is zero (negative label)
	'''
	from keras.layers import Input, Embedding, Reshape, Merge, Flatten, Activation
	from keras.models import Model

	word_input = Input(shape = (1,), dtype = 'int32')
	context_input = Input(shape = (num_ns+1,), dtype = 'int32')

	node_embedding = Embedding(num_nodes, embedding_dim)
	we = node_embedding(word_input)

	if seperate_context:
		context_embedding = Embedding(num_nodes, embedding_dim)
	else:
		context_embedding = node_embedding
	ce = Reshape((embedding_dim,num_ns+1))(context_embedding(context_input))

	dots = Flatten()(Merge(mode = 'dot', dot_axes = (1,2))([ce, we]))
	acts = Activation('sigmoid')(dots)
	
	model = Model(input = [word_input, context_input], output = acts)
	model.compile('adam', loss = 'binary_crossentropy')
```


3. Check out the keras.io documentation. The API supports "masking" which is what you're talking about.

**Comment by amityaffliction** [comment]
```
  model = Sequential()
  model.add(Embedding(1000, 64, input_length=10))
  # the model will take as input an integer matrix of size (batch, input_length).
  # the largest integer (i.e. word index) in the input should be no larger than 999 (vocabulary size).
  # now model.output_shape == (None, 10, 64), where None is the batch dimension.

  input_array = np.random.randint(1000, size=(32, 10))

  model.compile('rmsprop', 'mse')
  output_array = model.predict(input_array)
  assert output_array.shape == (32, 10, 64)

```
Above is Keras Embeddings layer example in the web page
http://keras.io/layers/embeddings/
It passes 'mse' as argument. I was asking how does embeddings layer do task

1. 'indice -> vector'
2. and what does objective function do in this process??

@jmhessel Thank you for your answer

**Comment by katomaso** [comment]
Quick look into the source code reveals that the Embedding layer has a set of (trainable) weights which are used as `params` in keras' engine method `gather`. Suppose tensorflow as the engine now.
https://www.tensorflow.org/versions/r0.9/api_docs/python/array_ops.html#gather

As you can see the result of Embedding is just a permutation of inner trainable weights. Where the permutation is denoted by "indices" in your input array.

**Comment by Capable-Boys** [comment]
Hello jmhessel and katomaso,

Thanks for your helpful guide.

Since in here you talked about how to deal with word embeddings especially in the skipgram model, I wanted to know how we can save an embedding layer in the way that can be seen in a regular word embeddings file (i.e. a text file with .txt extension). Let’s assume we either learn these word embeddings in the model from scratch or we update those pre-trained ones which are fed into the first layer of the current model. Is there any way around this?!

Thank you in advance.

**Comment by jmhessel** [comment]
If you've trained an embedding layer `emb`, you can get the resulting word-by-dimension matrix by `my_embeddings = emb.get_weights()`. Then, you can do normal numpy things like `np.save("my_embeddings.npy", my_matrix)`

**Comment by Capable-Boys** [comment]
Thanks jmhessel. Your previous answer was very helpful. 
However I have another question here. Let's assume we have built two columns of networks in keras and these two columns are exactly the same which will merge together on their top and then feed into a dense layer which is the output layer in the model. My question is, while the first layer of each column here is an embedding layer, how can we share the the weights of the similar layers in the columns? No need to say that we have a parameter (like 'separate_context' that you had in your code) set as false meaning we would only have one embedding matrix which we work with.
Thanks in advance.

**Comment by Tachyon5** [comment]
I am very curious how the embedding layer is different from skip-gram or other word2vec variants? Is it the same thing or different and if different, then how? Can I take any set of sparsely-coded categorical variables and embed them? I would love to see a paper but I only see a link to the "A Theoretically Grounded Application of Dropout in Recurrent Neural Networks" within the documents.

**Comment by Capable-Boys** [comment]
Skip-gram, CBOW, and GloVe (or any other word2vec variant) are pre-trained word embeddings that can be set as the weight of an embedding layer. If the weight of this layer (generally the first layer of the network) is not initialized by these pre-trained vectors, the model/network itself would assign random weights and will learn the embeddings (i.e. weights) on the fly. Also, when the embedding layer is benefiting from these pre-trained word embeddings, you can define whether you like the network update these weight (i.e. embeddings) or you would like to have them frozen (It is task dependent, but in general letting the network update the weights would give you a better answer).
I believe, in a naive way, you can consider any type of low-dimension data (which represent a high-dimension data) as your embeddings. Keep in mind that embeddings are not only limited to natural language words as in the recent studies they have applied it to the biological concepts and social graphs as well.

**Comment by einareinarsson** [comment]
The essence is still not have been explained, namely, how does the Embedding layer figure out the appropriate weights in the proper way (representing semantic relationship in the vector space)?

**Comment by PaulX-CN** [comment]
I have a question, when I am training with questions and answers pairs, I always need to reduce the size of my vocabulary down to say 10k. The rest words will become UNKNOWN in my vocab. Plus, there will be EOS and PAD in the sequence to make up a fixed length sequence. If I want to use pretrained embedding, what is the embedding for EOS and UNK then? I know that I can mask the PAD (by setting masking=True in embedding layer) so they are not considered a word but EOS and UNK should be treated as a meaning word, am I correct? Then what is the embedding weight for them? They do not exist in any pretrained model.

**Comment by Capable-Boys** [comment]
You need to have a method to construct distributional representations for <UNK> token. There are three strategies around this:
**onehot:** represent unknown token with a one-hot vector;
**averaged:** representation vector of unknown token is set to the vector of a word with frequency 1 that is the closest to an average of vectors of words with frequency 1;
**random:** representation vector of unknown token is set to a representation of a random word with frequency 1.

For the end of sentence token (usually represented as <\S>), which can be important sometimes, you can either look for a better pre-trained model that includes EOS, or you can use onehot strategy here as well.

**Comment by vsoto** [comment]
Is there a paper/citation for how the embedding layers work in keras?

**Comment by graydan** [comment]
So the embedding layer updates on the fly during training, and it performs a dictionary lookup, but how? Is it an RNN? I read "A Theoretically Grounded Application of Dropout in Recurrent Neural Networks," but its not clear how the the Keras embedding layer is implemented. Also, any help on where to start understanding the Keras source code would be great!

**Comment by zzkszzks** [comment]
According to keras's official example: imdb_fasttext.py, I think the embedding layer is a fasttext implement.

**Comment by zuxfoucault** [comment]
I'm not sure if my understanding is correct but...

One of the purpose I want to initiated a pre-trained fasttext weight in the embedding layers is to decrease the unknown words in the test environment (these unknown words are not in training set). Since pre-trained fasttext model has larger vocabulary, during test environment, the unknown word can be represented by fasttext out-of-vocabulary word vectors, which supposed to share similar direction of the semantic similar words in the training set.  

However, due to the fact that the initial fasttext weights in the embedding layers will be updated through the training process (as mentioned in previous discussion, this setting generates better result). I am wondering if the updated embedding weights would distort the relationship of semantic similarity between words and undermine the representation of fasttext out-of-vocabulary word vectors?

**Comment by fqassemi** [comment]
I have worked on NLP for a while, now. I understand tf-idf, and I do understand the word frequency. GloV and other word2vec are also understood (which transform words to vectors not to an integer). Now, converting a word to an integer, is still a myth for me! If it is just word_frequency, then it is fine (not very exciting though). I also have problem with "learning on the fly", Embedding is actually representation of our input, so input can not be learnt on the fly! (My point is that optimization for embedding layer should be done before neural net start training. Otherwise what we are learning). So, if anyone can just explain this with examples, just compare output results of embedding and GloV, or word2vec, and explain how a vector or a word is converted to an integer is really appreciated.

**Comment by ahmadpgh** [comment]
GloVe and word2vec DO NOT convert a word to an integer. They change the discrete/atomic representation of the vocabulary words from one-hot encoding representations (disjoint representation of words) to low-dimensional continuous/distributed vectors that we call word embeddings.

It is common to have these words embeddings computed using an unsupervised technique like GloVe or word2vec (or even other simpler techniques like LSI/LSA applied to the TF-IDF matrix).

In supervised applications, these embeddings vectors tend to sit on top of the input layer; this assignment will set the first weights of the supervised network. The input layer will be receiving the one-hot vectors of the (input) words and NOT their embeddings. During the training and backpropagation, the input of the network will be unchanged, that is to say, only these embedding weights (and other weights of the network) will be trained. Even though in practice it is common to initialize these embedding weights with the GloVe or word2vec embeddings attempting to put the network in a proper state prior to training; these weights can be assigned randomly and be trained during the supervised training of the model as well (not the best recommendation though, unless enough training data is available).

To make it more clear, assume we have this set of vocabulary words in the language/application: {'_bar_', '_foo_', '_baz_', '_qux_'}. For each of them we are going to have one one-hot vector like this:

One-hot vectors:
--------------------------
'_bar_': [1, 0, 0, 0]
'_foo_': [0, 1, 0, 0]
'_baz_': [0, 0, 1, 0]
'_quz_': [0, 0, 0, 1],

Also, we can have for each of these words a word embedding (i.e. a word vector) of size 2 (one of the hyperparameters of the network - typically a number between 50 to 300) pre-trained or randomly initialize:

Word (embedding) vectors:
---------------------------------------
'_bar_': [+0.10, +0.25]
'_foo_': [-0.23, +0.61]
'_baz_': [+0.00, -0.05]
'_quz_': [-0.11, -0.17]

If you look at the above vectors stacked on top off each other as they are shown, they form a matrix. Now, we have our lookup table (the above matrix) from which the embeddings come and in which the embeddings will be updated. 

Let's imagine the input of the network is the sentence "_quz foo_" (you can look at it as a sequence or BoWs). The vertical one-hot vectors of '_quz_' and 'foo' will be multiplied to the lookup table to **_on_** their associated embeddings accordingly and _**off**_ the embeddings of the absent words. The result of this multiplication will be the input of the next layer in the network (can be feedforward, CNN, LSTM, etc). And, during the backpropagation, only these weights (i.e. embeddings) will be trained (if we set the network to allow this update).

FYI, in practice, for computational efficiency, there is no multiplication of one-hot vectors to the lookup table. I.e, for each word, one index is assigned which is equivalent to the row number of that word in the lookup table. Knowing this index number is going to help to retrieve the associated word vector of the given word from the lookup table much faster.

**Comment by fqassemi** [comment]
I agree, as I also mentioned word2vec type embedding represent a word as a vector. Just to summarize what I understood, an embedding in keras works like an extra layer at the interface of words and the rest of layers, where the matrix coefficient (W_ij) is just given by a pre-trained model such as GloVe.  This extra layer will act as an input to the NN (that is, input_dim is equal to embedding_dim no matter the number of words in the text). If so, then Embedding Layer makes all sense to me.

**Comment by ahmadpgh** [comment]
@fqassemi You are very welcome. I fully updated my answer above. In our toy example, the _input_dim_ is 5 (i.e. 4+1) and the _embedding_dim_ is 2.
